src/main/java/net/glowstone/block/blocktype/BlockPiston.java

```diff
@@ -158,6 +216,92 @@ public void onRedstoneUpdate(GlowBlock me) {
         setType(me.getRelative(pistonBlockFace), 0, 0);
     }
 
+    // FIXME Remove this
+    private static Logger logger = Logger.getLogger(BlockPiston.class.getName());
+
+    private static final BlockFace[] ADJACENT_FACES = new BlockFace[] {BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST, BlockFace.UP, BlockFace.DOWN};
+
+    private boolean addBlock(GlowBlock piston, BlockFace movementDirection,
+                             BlockInfo blockInfo, List<Block> blocksToMove, List<Block> blocksToBreak) {
+
+        GlowBlock block = blockInfo.getBlock();
+        BlockFace ignoredFace = blockInfo.getIgnoredFace();
+        boolean isPushing = (movementDirection == ignoredFace.getOppositeFace());
+        MaterialValueManager.ValueCollection materialValues = block.getMaterialValues();
+        PistonMoveBehavior moveBehavior =
+            isPushing ? materialValues.getPistonPushBehavior() : materialValues.getPistonPullBehavior();
+
+        if (block.isEmpty()) {
+            return true;
+        }
+
+        if (blocksToMove.size() >= PUSH_LIMIT) {
+            return false;
+        }
+
+        if (isPushing) {
+            switch (moveBehavior) {
+                case MOVE:
+                    blocksToMove.add(block);
+                    break;
+                case BREAK:
+                    blocksToBreak.add(block);
+                    return true;
+                case DONT_MOVE:
+                    return false;
+            }
+        } else {
+            switch (moveBehavior) {
+                case MOVE:
+                    blocksToMove.add(block);
+                    break;
+                case BREAK:
+                case DONT_MOVE:
+                    return true;
+            }
+        }
+
+        if (block.getType() == Material.SLIME_BLOCK) {
```

Comment:
Extract this into a subroutine. Add a PistonMoveBehavior.STICKY_MOVE and make it the value for Material.SLIME_BLOCK instead of checking for SLIME_BLOCK directly; that way, we can add the behavior to other blocks in the future.